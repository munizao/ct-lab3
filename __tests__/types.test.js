const {
  isNumber,
  isString,
  isBoolean,
  isArray,
  isObject,
  isFunction,
  castToNumber,
  castToString,
  castToBoolean,
  castToArray,
  castToObject,
  castToFunction,
  getCaster
} = require('../lib/types.js');

describe('validator module', () => {
  describe('basic validation', () => {
    it('properly tells if a value is a number', () => {
      expect(isNumber(3)).toBeTruthy();
      expect(isNumber('hi')).toBeFalsy();
      expect(isNumber([])).toBeFalsy();
      expect(isNumber({})).toBeFalsy();
      expect(isNumber(() => {})).toBeFalsy();
      expect(isNumber(true)).toBeFalsy();
    });
    it('properly tells if a value is a string', () => {
      expect(isString('hi')).toBeTruthy();
      expect(isString(3)).toBeFalsy();
      expect(isString([])).toBeFalsy();
      expect(isString({})).toBeFalsy();
      expect(isString(() => {})).toBeFalsy();
      expect(isString(true)).toBeFalsy();
    });
    it('properly tells if a value is a boolean', () => {
      expect(isBoolean(true)).toBeTruthy();
      expect(isBoolean('hi')).toBeFalsy();
      expect(isBoolean(3)).toBeFalsy();
      expect(isBoolean([])).toBeFalsy();
      expect(isBoolean({})).toBeFalsy();
      expect(isBoolean(() => {})).toBeFalsy();
    });
    it('properly tells if a value is an array', () => {
      expect(isArray([])).toBeTruthy();
      expect(isArray(true)).toBeFalsy();
      expect(isArray('hi')).toBeFalsy();
      expect(isArray(3)).toBeFalsy();
      expect(isArray({})).toBeFalsy();
      expect(isArray(() => {})).toBeFalsy();
    });
    it('properly tells if a value is an object', () => {
      expect(isObject({})).toBeTruthy();
      expect(isObject(null)).toBeFalsy();
      expect(isObject([])).toBeFalsy();
      expect(isObject(true)).toBeFalsy();
      expect(isObject('hi')).toBeFalsy();
      expect(isObject(3)).toBeFalsy();
      expect(isObject(() => {})).toBeFalsy();
    });
    it('properly tells if a value is a function', () => {
      expect(isFunction(() => {})).toBeTruthy();
      expect(isFunction({})).toBeFalsy();
      expect(isFunction(null)).toBeFalsy();
      expect(isFunction([])).toBeFalsy();
      expect(isFunction(true)).toBeFalsy();
      expect(isFunction('hi')).toBeFalsy();
      expect(isFunction(3)).toBeFalsy();
    });

  });

  describe('casters', () => {
    it('can cast values to a number', () => {
      expect(castToNumber(3)).toEqual(3);
      expect(castToNumber('3')).toEqual(3);
      expect(castToNumber(true)).toEqual(1);
      expect(castToNumber(false)).toEqual(0);
    });

    it('throws error if value is not castable to number', () => {
      expect(() => castToNumber('hi')).toThrowErrorMatchingSnapshot();
      expect(() => castToNumber('')).toThrowErrorMatchingSnapshot();
      expect(() => castToNumber({})).toThrowErrorMatchingSnapshot();
    });

    it('can cast values to a string', () => {
      expect(castToString(3)).toEqual('3');
      expect(castToString(true)).toEqual('true');
      expect(castToString([1, 2, 3, 4])).toEqual('1,2,3,4');
      expect(castToString({})).toEqual('[object Object]');
    });

    it('can cast values to a boolean', () => {
      expect(castToBoolean(true)).toEqual(true);
      expect(castToBoolean(3)).toEqual(true);
      expect(castToBoolean(0)).toEqual(false);
      expect(castToBoolean([1, 2, 3, 4])).toEqual(true);
      expect(castToBoolean({})).toEqual(true);
    });

    it('can cast values to an array', () => {
      expect(castToArray([1, 2])).toEqual([1, 2]);
      expect(castToArray('cat')).toEqual(['c', 'a', 't']);
      expect(castToArray({ color:'blue', isLarge: true })).toEqual([['color', 'blue'], ['isLarge', true]]);
    });

    it('throws error if value is not castable to array', () => {
      expect(() => castToArray(3)).toThrowErrorMatchingSnapshot();
      expect(() => castToArray(true)).toThrowErrorMatchingSnapshot();
    });

    it('can cast values to an object', () => {
      expect(castToObject({ color: 'blue' })).toEqual({ color: 'blue' });
      expect(castToObject([['color', 'blue'], ['isLarge', true]])).toEqual({ color:'blue', isLarge: true });
    });

    it('throws error if value is not castable to object', () => {
      expect(() => castToObject(3)).toThrowErrorMatchingSnapshot();
      expect(() => castToObject(true)).toThrowErrorMatchingSnapshot();
    });

    it('can cast values to a function', () => {
      const func = () => {};
      expect(castToFunction(func)).toEqual(func);
    });

    it('throws error if value is not castable to function', () => {
      expect(() => castToFunction(3)).toThrowErrorMatchingSnapshot();
      expect(() => castToFunction(true)).toThrowErrorMatchingSnapshot();
    });
  });

  it('can get the right caster', () => {
    expect(getCaster(Number)).toEqual(castToNumber);
    expect(getCaster(String)).toEqual(castToString);
    expect(getCaster(Boolean)).toEqual(castToBoolean);
    expect(getCaster(Array)).toEqual(castToArray);
    expect(getCaster(Object)).toEqual(castToObject);
    expect(getCaster(Function)).toEqual(castToFunction);
    expect(getCaster(Promise)).toBeNull();
  });
});
