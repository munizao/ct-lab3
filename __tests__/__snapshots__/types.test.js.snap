// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validator module casters throws error if value is not castable to array 1`] = `"Cannot cast >>3<< to Array"`;

exports[`validator module casters throws error if value is not castable to array 2`] = `"Cannot cast >>true<< to Array"`;

exports[`validator module casters throws error if value is not castable to function 1`] = `"Cannot cast >>3<< to Function"`;

exports[`validator module casters throws error if value is not castable to function 2`] = `"Cannot cast >>true<< to Function"`;

exports[`validator module casters throws error if value is not castable to number 1`] = `"Cannot cast >>hi<< to Number"`;

exports[`validator module casters throws error if value is not castable to number 2`] = `"Cannot cast >><< to Number"`;

exports[`validator module casters throws error if value is not castable to number 3`] = `"Cannot cast >>[object Object]<< to Number"`;

exports[`validator module casters throws error if value is not castable to object 1`] = `"Cannot cast >>3<< to Object"`;

exports[`validator module casters throws error if value is not castable to object 2`] = `"Cannot cast >>true<< to Object"`;
